pandas/core/reshape/concat.py

```diff
@@ -585,6 +625,66 @@ def _maybe_check_integrity(self, concat_index: Index):
                     "{overlap!s}".format(overlap=overlap)
                 )
 
+    def _items_overlap_with_suffix(self, objs, suffixes):
+        """
+        Adding suffix for items if there is overlapping situation.
+
+        Be aware that `objs` can be either DataFrame-like or Index-like given
+        if `self._is_series` is True or False.
+        """
+        if self._is_series:
+
+            # when _is_series is True, objs are actually column Index
+            overlap_cols = list(objs)
+        else:
+            overlap_cols = chain.from_iterable([obj.columns for obj in objs])
+        to_rename = [col for col, cnt in Counter(overlap_cols).items() if cnt > 1]
+
+        if len(to_rename) == 0 or suffixes is None:
+            return objs
+
+        if not isinstance(suffixes, tuple):
+            raise ValueError(
+                f"Invalid type {type(suffixes)} is assigned to suffixes, only "
+                f"'tuple' is allowed."
+            )
+
+        if len(objs) != len(suffixes):
+            raise ValueError(
+                "Number of objects for concatenation is not"
+                "equal to number of suffixes"
+            )
+
+        def renamer(x, suffix):
+            """
+            Rename the indices.
+
+            If there is overlap, and suffix is not None, add
+            suffix, otherwise, leave it as-is.
+
+            Parameters
+            ----------
+            x : original column name
+            suffix : str or None
+
+            Returns
+            -------
+            x : renamed column name
+            """
+            if x in to_rename and suffix is not None:
+                return f"{x}{suffix}"
+            return x
+
+        if self._is_series:
+            new_cols = [renamer(obj, suffix) for obj, suffix in zip(objs, suffixes)]
+            return new_cols
+
+        for obj, suffix in zip(objs, suffixes):
+            col_renamer = partial(renamer, suffix=suffix)
+            obj.columns = _transform_index(obj.columns, col_renamer)
```

Comment:
I think this line is causing the original dataframes to be modified as well:
```python
In [1]: import pandas as pd; pd.__version__
Out[1]: '0.26.0.dev0+947.gc8570707c'

In [2]: df1 = pd.DataFrame({'A': list('ab'), 'B': [0, 1]})

In [3]: df2 = pd.DataFrame({'A':list('ac'), 'C': [100, 200]})

In [4]: df3 = pd.concat([df1, df2], axis=1, suffixes=('_x', '_y'))

In [5]: df1.columns
Out[5]: Index(['A_x', 'B'], dtype='object')

In [6]: df1
Out[6]: 
  A_x  B
0   a  0
1   b  1

In [7]: df2.columns
Out[7]: Index(['A_y', 'C'], dtype='object')

In [8]: df2
Out[8]: 
  A_y    C
0   a  100
1   c  200
```
Root cause could be occurring elsewhere though, but it's an issue nonetheless.  Would be nice to add a test for the above.